Use MSBuild Static Graph to evaluate projects

jeffkl opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/8791
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: This change adopts Static Graph to speed up command-line based restores.  There is a new EXE that does the evaluation and restore out of proc so that server GC can be enabled.  The feature is off by default and users must set a property to enable it.

## Testing/Validation

Tests Added: Yes/No  
Reason for not adding tests:  
Validation:

jeffkl commented:
/cc @AndyGerlicher

dtivel commented on NuGet.sln:
```diff
@@ -169,7 +168,6 @@ EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "NuGet.VisualStudio.Interop", "src\NuGet.Clients\NuGet.VisualStudio.Interop\NuGet.VisualStudio.Interop.csproj", "{7DB43FE1-75E1-49F9-B2C8-06A552BA2144}"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NuGetConsole.Host.PowerShell", "src\NuGet.Clients\NuGetConsole.Host.PowerShell\NuGetConsole.Host.PowerShell.csproj", "{5A79EEF3-51C0-4A14-8D37-50EF38AD835D}"
-	# Make NuGetConsole.Host.PowerShell depend on NuGet.PackageManagement.PowerShellCmdlets.  This is so NuGetConsole.Host.PowerShell's build can update NuGet.psd1 based on the strong name of NuGet.PackageManagement.PowerShellCmdlets.dll
```
Was this removal intentional?  If yes, why?

jeffkl replied:
Visual Studio did this when I added the new project via the UI.  Should I bring the comment back?

dtivel replied:
Yes, please.

dtivel commented on build/config.props:
```diff
@@ -54,7 +54,7 @@
     <MoqVersion>4.12.0</MoqVersion>
     <FluentAssertionsVersion>4.19.4</FluentAssertionsVersion>
-    <MicrosoftBuildPackageVersion Condition="$(MicrosoftBuildPackageVersion) == ''">16.0.461</MicrosoftBuildPackageVersion>
+    <MicrosoftBuildPackageVersion Condition="$(MicrosoftBuildPackageVersion) == ''">16.4.0-preview-19517-01</MicrosoftBuildPackageVersion>
```
What feed does this come from?  The most recent package version I see on https://dotnet.myget.org/feed/msbuild/package/nuget/Microsoft.Build.Utilities.Core is 16.4.0-preview-19516-02.

jeffkl replied:
I think MSBuild is now pushing packages to https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json instead

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs:
```diff
@@ -0,0 +1,47 @@
+using System;
```
Add a license header like [here](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-4530787af238fb65078a3652ea65a56fR1-R2).

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,173 @@
+using System;
```
Add a license header.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,173 @@
+
+                Log.LogMessage(MessageImportance.Low, "\"{0}\" {1}", startInfo.FileName, startInfo.Arguments);
+
+                using (var semaphore = new SemaphoreSlim(0, 1))
```
Use named parameters for literal arguments.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,173 @@
+            }
+        }
+
+        private Dictionary<string, string> GetProperties()
```
Delete this unreferenced method.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,173 @@
+            return !Log.HasLoggedErrors;
+        }
+
+        private void AddPropertyIfTrue(Dictionary<string, string> properties, string name, bool value)
```
Can be `static`.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,173 @@
+#endif
+                    CreateNoWindow = true,
+#if !NETFRAMEWORK
+                    FileName = Path.GetFullPath(Path.Combine(MSBuildBinPath, @"..", "..", "dotnet")),
```
Remove `@`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,173 @@
+
+                var startInfo = new ProcessStartInfo
+                {
+#if !NETFRAMEWORK
```
Invert logic to improve readability here and below:
```C#
#if NETFRAMEWORK
    ...
#else
    ...
#endif
```

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,173 @@
+
+namespace NuGet.Build.Tasks
+{
+    public class RestoreTaskEx : Task, ICancelableTask, IDisposable
```
`sealed`.  It simplifies the `IDisposable` implementation.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,303 @@
+
+using System;
+using System.Collections;
+using System.Threading.Tasks;
```
Remove.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,102 @@
+
+        public static async Task<int> Main(string[] args)
+        {
+            var properties = ParseProperties(args[0]);
```
There should be some basic argument checking.  If <4 arguments are passed the process should not crash with an unhandled exception.

jeffkl replied:
Technically the only thing that launches this is NuGet so I wasn't sure how much checking to to do?  I guess users might get sneaky and try to execute it...

dtivel replied:
For starters, how about avoiding the unhandled exception?
```C#
private const int Succeeded = 0;
private const int Failed = 1;

...

if (args.Length != 4)
{
    return Task.FromResult(Failed);
}

…

return await dependencyGraphSpecGenerator.RestoreAsync(properties) ? Succeeded : Failed;
```

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,102 @@
+            var msbuildExePath = new FileInfo(args[1]);
+            var entryProjectPath = args[2];
+            var globalProperties = ParseProperties(args[3]);
+            var debug = globalProperties.TryGetValue("NuGetDebug", out var debugValue) && string.Equals(debugValue, "true", StringComparison.OrdinalIgnoreCase);
```
We don't have a single environment variable to debug all things NuGet.  This sounds like that.

For now I'd prefer to follow the prevalent environment variable naming pattern and name this something more specific like `NUGET_RESTORE_DEBUG`.

jeffkl replied:
Thanks for commenting, I meant to leave a `TODO` here since I wasn't in love with the name either.  Let's discuss in the review meeting

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs:
```diff
@@ -0,0 +1,47 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Remove.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs:
```diff
@@ -0,0 +1,47 @@
+            return IsValueTrue(item.GetMetadataValue(name), defaultValue);
+        }
+
+        public static bool IsPropertyTrue(this ProjectInstance projectInstance, string name, bool defaultValue = false)
```
Nowhere is `true` passed for `defaultValue` so remove the parameter.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,303 @@
+
+namespace NuGet.Build.Tasks
+{
+    internal enum ConsoleOutLogMessageType
```
Put in separate file.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,303 @@
+                }
+
+
+                if (message.Length >= 2 && message[0] == '{' && message[message.Length - 1] == '}')
```
This can be moved outside of the lock.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,303 @@
+                        default:
+                            throw new ArgumentOutOfRangeException();
+                    }
+                } /**/
```
Remove comment.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,303 @@
+                    }
+                    catch
+                    {
+                        log.LogMessage(null, null, null, 0, 0, 0, 0, message, null, null, MessageImportance.Low);
```
Add named parameters.

dtivel commented on src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj:
```diff
@@ -80,6 +79,10 @@
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildPackageVersion)" />
```
For this and the [above change](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-94c7bd5ae06e284fdedbed8652a0d7e0L62), what about [NuGet.PackageManagement.VisualStudio](https://github.com/jeffkl/NuGet.Client/blob/63c2f47fa67e0e6ee0f347e4f366b6f65fff952b/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/NuGet.PackageManagement.VisualStudio.csproj#L54), [NuGet.Build.Tasks.Pack](https://github.com/jeffkl/NuGet.Client/blob/63c2f47fa67e0e6ee0f347e4f366b6f65fff952b/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj#L37-L46), and [NuGet.Build.Tasks.Pack.Test](https://github.com/jeffkl/NuGet.Client/blob/63c2f47fa67e0e6ee0f347e4f366b6f65fff952b/test/NuGet.Core.Tests/NuGet.Build.Tasks.Pack.Test/NuGet.Build.Tasks.Pack.Test.csproj#L24-L31)?

jeffkl replied:
Good question.  Let's discuss this one in the review meeting

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/NuGet.Build.Tasks.Console.csproj:
```diff
@@ -0,0 +1,43 @@
+  <PropertyGroup>
+    <TargetFrameworks>$(TargetFrameworksExe)</TargetFrameworks>
+    <OutputType>Exe</OutputType>
+    <TargetFramework />
```
Move up closer to its friend `<TargetFrameworks />` like [here](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-148a5d9a5d8aec6948f9ead08e0f5246R21-R22).

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/NuGet.Build.Tasks.Console.csproj:
```diff
@@ -0,0 +1,43 @@
+    <NoWarn>$(NoWarn);CS1591</NoWarn>
+    <XPLATProject>true</XPLATProject>
+    <EnableDefaultNoneItems>false</EnableDefaultNoneItems>
+    <ServerGarbageCollection>true</ServerGarbageCollection>
```
What does this do?  Where is it used?

jeffkl replied:
It tells .NET Core to enable server GC via the application's `runtimeconfig.json` which looks like this:
```diff
{
  "runtimeOptions": {
    "tfm": "netcoreapp2.1",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "2.1.0"
    },
    "configProperties": {
+      "System.GC.Server": true
    }
  }
}
```

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,102 @@
+    {
+        private static readonly string[] DefaultWildcardMatchesToSkip =
+        {
+            // this is an ends with **\* without any *s.
```
...any _other_ *s?

jeffkl replied:
It only matches `**\*` and nothing else like `**\**`

jeffkl replied:
This is now a whitelist essentially, any wildcards and does not end in `proj`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,102 @@
+            Environment.SetEnvironmentVariable("MSBUILD_EXE_PATH", msbuildExePath.FullName);
+            Environment.SetEnvironmentVariable("MSBuildCacheFileEnumerations", "1");
+            Environment.SetEnvironmentVariable("MSBuildLoadAllFilesAsReadonly", "1");
+            Environment.SetEnvironmentVariable("MSBuildSkipEagerWildCardEvaluationRegexes", string.Join(";", DefaultWildcardMatchesToSkip));
```
Where are these variables documented?  It would be good to have brief explanations for why these options need to be set as they are.

jeffkl replied:
https://github.com/microsoft/msbuild/blob/master/src/Shared/Traits.cs

I'll add some comments in this code to explain a little as well

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,102 @@
+
+                var path = Path.Combine(msbuildDirectory, $"{assemblyName.Name}.dll");
+
+                return File.Exists(path) ? Assembly.LoadFrom(path) : null;
```
This precludes native image loading.  Do we have to?

jeffkl replied:
Since this is running as a different process and its assembly is next to NuGet, the MSBuild assemblies won't be found when running in full framework.  .NET Core won't do this since NuGet and MSBuild are in the same folder.

jeffkl replied:
We looked into this and using an entry in the app.config is the better option:

```xml
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="Microsoft.Build.Framework" culture="neutral" publicKeyToken="b03f5f7f11d50a3a" />
        <bindingRedirect oldVersion="0.0.0.0-99.9.9.9" newVersion="15.1.0.0" />
        <codeBase version="15.1.0.0" href="..\..\..\..\..\MSBuild\Current\Bin\Microsoft.Build.Framework.dll" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
</configuration>
```

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,303 @@
+
+        public static void LogToConsole(BuildMessageEventArgs buildMessageEventArgs)
+        {
+            lock (LockObject)
```
Why not move the locking into [`LogToConsole(ConsoleOutLogMessage)`](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-e42d9ac738e39577a5d4d8a3a3b937e4R178-R181)?  Then you only need to lock in one place.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,303 @@
+                }
+            }
+
+            public void Initialize(IEventSource eventSource)
```
Handle `null` `eventSource`?

jeffkl replied:
MSBuild calls this method and will always supply a value

dtivel reviewed (commented):
I haven't finished reviewing this but posting comments so far anyway.  Some may be outdated.

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+
+        public DependencyGraphSpecGenerator(string entryProjectPath, Dictionary<string, string> globalProperties)
+        {
+            _entryProjectPath = entryProjectPath;
```
Do we need argument null checks?

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+
+        private readonly int _loggerVerbosity;
+
+        public DependencyGraphSpecGenerator(string entryProjectPath, Dictionary<string, string> globalProperties)
```
Could we have comments? What type of elements are expected in the dictionary?

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+
+        public bool Debug { get; set; }
+
+        public async Task<bool> RestoreAsync(Dictionary<string, string> properties)
```
Comments?

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+                loggers: null,
+                remoteLoggers: null,
+                toolsetDefinitionLocations: ToolsetDefinitionLocations.Default,
+                maxNodeCount: 1,
```
Why 1? Does it matter if more ?

jeffkl replied:
I'll add a comment

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+                                            failedBuildSubmissions.Add(buildSubmission);
+                                        }
+                                    },
+                                    context: null);
```
Will these calls be awaited? Is the line BuildManager.DefaultBuildManager.EndBuild(); waiting for them?

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+                                        TargetsToBuild,
+                                        hostServices: null,
+                                        BuildRequestDataFlags.SkipNonexistentTargets))
+                                .ExecuteAsync(
```
What happens on these calls? Will the projects be built?

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+
+                        var projectInstance = project.CreateProjectInstance(ProjectInstanceSettings.None, evaluationContext);
+
+                        if (properties.ContainsKey("TargetFramework") || string.IsNullOrWhiteSpace(projectInstance.GetPropertyValue("TargetFrameworks")))
```
Why this branch? Will the legacy projects be supported?

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+namespace NuGet.Build.Tasks.Console
+{
+    internal class DependencyGraphSpecGenerator
+    {
```
I did not see tests for the new types.

jeffkl replied:
I will be adding unit tests in a future PR

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,222 @@
+#endif
+                    CreateNoWindow = true,
+#if NETFRAMEWORK
+                    FileName = Path.Combine(thisAssembly.DirectoryName, Path.ChangeExtension(thisAssembly.Name, ".Console.exe")),
```
The two assemblies(NuGet.Build.Tasks and NuGet.Build.Tasks.Console( need to stay in syn regarding with the name. Do we need an unit test?

cristinamanum replied:
Why does the execution needs to be out of proc?

cristinamanum commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,802 @@
+
+        private static List<LibraryDependency> GetPackageReferences(ProjectInstance projectInstance)
+        {
+            var packageReferenceItems = projectInstance.GetItems("PackageReference").Distinct(ProjectItemInstanceEvaluatedIncludeComparer.Instance).ToList();
```
Are two PackageReferences same Id but different metadata equal?

cristinamanum replied:
Is it expected that the target CollectPackageReferences was executed before this call?

jeffkl committed:
Use MSBuild Static Graph to evaluate projects

This change adopts Static Graph to speed up command-line based restores.  There is a new EXE that does the evaluation and restore out of proc so that server GC can be enabled.

jeffkl commented:
This is now ready for a more final review.  I have merged a few changes to reduce code duplication and have updated setup logic so that the new EXE is distributed in the VSIX.

I'll need to rebase this on Kat's change that adds support for packages.config and then I'll have a final review.

jeffkl committed:
Add option for configuring a binary logging

jeffkl committed:
Pass along global properties

jeffkl committed:
Update pack test

jeffkl committed:
Add some unit tests

jeffkl committed:
Include System.Threading.Tasks.Dataflow for ILMerge

jeffkl committed:
Port packages.config changes

jeffkl committed:
Update unit test for alternate directory separators

jeffkl commented:
This is ready for final review and approval.  I've ported the packages.config changes and have a passing CI.  All that's left is to add some end-to-end tests which I'll do later after I've done a VS and .NET Core insertion.

dtivel commented on build/config.props:
```diff
@@ -54,7 +54,7 @@
     <MoqVersion>4.12.0</MoqVersion>
     <FluentAssertionsVersion>4.19.4</FluentAssertionsVersion>
-    <MicrosoftBuildPackageVersion Condition="$(MicrosoftBuildPackageVersion) == ''">16.0.461</MicrosoftBuildPackageVersion>
+    <MicrosoftBuildPackageVersion Condition="$(MicrosoftBuildPackageVersion) == ''">16.5.0-preview-19606-01</MicrosoftBuildPackageVersion>
```
Should `$(MicrosoftBuildPackageVersion)` be in single quotes?
```XML
Condition="'$(MicrosoftBuildPackageVersion)' == ''"
```

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/App.config:
```diff
@@ -0,0 +1,62 @@
+      <dependentAssembly>
+        <assemblyIdentity name="System.Numerics.Vectors" culture="neutral" publicKeyToken="b03f5f7f11d50a3a" />
+        <codeBase version="4.1.3.0" href="..\..\..\..\..\MSBuild\Current\Bin\System.Numerics.Vectors.dll" />
+      </dependentAssembly>
```
If MSBuild updates any of these assemblies, how fragile will this be?  How do we stay in sync?

jeffkl replied:
@rainersigwald I'm taking a dependency on the versions of these assemblies that MSBuild is shipping.  Is there a good way to keep these in sync?

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Client/NuGet.VisualStudio.Client.csproj:
```diff
@@ -111,6 +115,14 @@
+      <Ngen>True</Ngen>
+      <NgenApplication>$(NgenApplicationDefault)</NgenApplication>
+      <NgenArchitecture>X86</NgenArchitecture>
+      <NgenPriority>$(NgenPriorityDefault)</NgenPriority>
```
NGEN-ing NuGet.Build.Tasks.Console.exe seems like a good idea, but it would be good to get some idea of how impactful NGEN-ing would be.  This won't work as-is; you'd need to add a new test and more configuration to enable NGEN-ing.

I'd remove this for now.  We can discuss later.

dtivel replied:
Remove lines 121-124 above.
```XML
      <Ngen>True</Ngen> 
      <NgenApplication>$(NgenApplicationDefault)</NgenApplication> 
      <NgenArchitecture>X86</NgenArchitecture> 
      <NgenPriority>$(NgenPriorityDefault)</NgenPriority>
```

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs:
```diff
@@ -0,0 +1,194 @@
+                        break;
+
+                    default:
+                        throw new ArgumentOutOfRangeException(nameof(value), value, null);
```
```C#
message: null
```

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs:
```diff
@@ -0,0 +1,194 @@
+        /// <inheritdoc cref="ILogger.Shutdown" />
+        void ILogger.Shutdown()
+        {
+            if (_eventSource != null)
```
Why guard against `null` here but not [there](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-155099ed29f34e5d459bd7cd8d4b5d1fR127-R131)?

jeffkl replied:
The `Initialize` method is called by MSBuild and `eventSource` is never null.  So if Initialize isn't called by MSBuild because of some exception, then `_eventSource` could be null.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildProjectItemInstance.cs:
```diff
@@ -0,0 +1,33 @@
+    /// <summary>
+    /// Represents an <see cref="MSBuildItemBase" /> implementation which wraps a <see cref="ProjectItemInstance" /> object.
+    /// </summary>
+    internal class MSBuildProjectItemInstance : MSBuildItemBase
```
Can be `sealed`?

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildItemBase.cs:
```diff
@@ -0,0 +1,51 @@
+
+            if (trim)
+            {
+                value = value.Trim();
```
Do you need to guard against `null`?

jeffkl replied:
No, the implementing classes call MSBuild APIs which always return `string.Empty` instead of `null`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/TaskLoggingQueue.cs:
```diff
@@ -0,0 +1,99 @@
+
+                try
+                {
+                    consoleOutLogMessage = JsonConvert.DeserializeObject<ConsoleOutLogMessage>(message);
```
Is there no performance benefit from using a custom `JsonSerializerSettings` (like [here](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-155099ed29f34e5d459bd7cd8d4b5d1fR33-R37)?

jeffkl replied:
I mainly wanted to improve serialization performance by leaving out null properties.  I wonder if a custom deserializer should be the best option.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/TaskLoggingQueue.cs:
```diff
@@ -0,0 +1,99 @@
+                catch
+                {
+                    // Log the raw message if it couldn't be deserialized
+                    _log.LogMessage(null, null, null, 0, 0, 0, 0, message, null, null, MessageImportance.Low);
```
I think the wrong overload is being called.  Reminder to use named parameters for all those mysterious literals.

In this case, I think you want:
```C#
_log.LogMessage(MessageImportance.Low, message);
```

You can add a unit test for this and the rest of the class.

jeffkl replied:
Done

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs:
```diff
@@ -0,0 +1,194 @@
+            get => _verbosity;
+            set
+            {
+                _verbosity = value;
```
This statement should come after the `switch` statement below.

Suppose `value` is an unrecognized `LoggerVerbosity` value; the current code would update `_verbosity` and then throw.

jeffkl replied:
Its not considered an exception to set an enum to a undefined value.  But I'll move the assignment down

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs:
```diff
@@ -0,0 +1,194 @@
+        }
+
+        /// <inheritdoc cref="IBuildEngine.LogCustomEvent" />
+        void IBuildEngine.LogCustomEvent(CustomBuildEventArgs e) { }
```
These events are uninteresting?  Or, they're never raised?

jeffkl replied:
I am not aware of a situation where they are raised.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs:
```diff
@@ -0,0 +1,194 @@
+        /// </summary>
+        /// <param name="sender">The sender of the event.</param>
+        /// <param name="e">The <see cref="BuildWarningEventArgs" /> containing the details of the message event.</param>
+        private void OnWarningRaised(object sender, BuildWarningEventArgs e) => Enqueue(e);
```
Warnings and errors are always logged, even with `LoggerVerbosity.Quiet`?

According to https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.framework.loggerverbosity?view=netframework-4.8#fields,
* It's unclear if `LoggerVerbosity.Quiet` logs warnings or errors, but it seems like it doesn't log any messages; all `LoggerVerbosity` levels except `LoggerVerbosity.Quiet` log `MessageImportance.High`.
* It seems only `LoggerVerbosity.Detailed` and `LoggerVerbosity.Diagnostic` log `MessageImportance.Normal` messages.

However, [here](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-155099ed29f34e5d459bd7cd8d4b5d1fR101-R113):
*  `LoggerVerbosity.Quiet` logs `MessageImportance.High`.
*  `LoggerVerbosity.Normal` logs `MessageImportance.Normal`.

jeffkl replied:
Yes, warnings and errors are always logged.  The verbosity setting is only for messages.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs:
```diff
@@ -0,0 +1,53 @@
+    /// <summary>
+    /// Represents extension methods to built-in types.
+    /// </summary>
+    internal static class ExtensionMethods
```
MSBuild assemblies don't already have public utility methods we can reuse?

jeffkl replied:
Unfortunately no.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/IMSBuildProject.cs:
```diff
@@ -0,0 +1,31 @@
+    /// <summary>
+    /// Represents an MSBuild project.
+    /// </summary>
+    internal interface IMSBuildProject : IMSBuildItem
```
Why does `IMSBuildProject` derive from `IMSBuildItem`?  It implies that [`GetItems(...)`](https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-3a9434041dbc16a778ee86149c32ece1R29) could theoretically return some instances of `IMSBuildProject`, which seems odd.

jeffkl replied:
I did this because IMSBuildItem has every method I needed except the project path and directory.  So I had IMSBuildProject derive from IMSBuildItem to add these two properties.  GetItems could return an IMSBuildProject but then you'd have to cast it since its returning IMSBuildItem.  These are all internal so I don't think it matters.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildFeatureFlags.cs:
```diff
@@ -0,0 +1,79 @@
+        /// to be found by MSBuild (stuff like $(MSBuildExtensionsPath).
+        /// More info here: https://github.com/microsoft/msbuild/blob/master/src/Shared/BuildEnvironmentHelper.cs#L125
+        /// </remarks>
+        public static string MSBuildExePath
```
`MSBuildExeFilePath`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,150 @@
+                System.Console.WriteLine("Waiting for debugger to attach to Process ID: {Process.GetCurrentProcess().Id}");
+
+                while (!Debugger.IsAttached)
+                    System.Threading.Thread.Sleep(100);
```
Use explicit {} block.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <returns><code>true</code> if the user specified to debug the current process, otherwise <code>false</code>.</returns>
+        private static bool IsDebug()
+        {
+            return string.Equals(Environment.GetEnvironmentVariable("DEBUG_RESTORE_TASK"), bool.TrueString, StringComparison.OrdinalIgnoreCase);
```
I would prefer that NuGet environment variables have a `NUGET_` prefix.  I understand that there may be some existing environment variables not following this pattern, but many do, and I feel it's a good pattern going forward.

jeffkl replied:
I opted to re-use an existing env var:

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks/RestoreTask.cs#L78

Should I rename both of them, have them be different, or support both?

dtivel replied:
Looks fine to me.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,150 @@
+                // Command-line is correct if no exceptions were thrown and the MSBuild path exists and an entry project were specified
+                return msbuildExePath.Exists && !string.IsNullOrWhiteSpace(entryProjectPath);
+            }
+            catch
```
```C#
catch (Exception)
```
...is preferable.  See the 4th paragraph [here](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/try-catch).

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <param name="args">A <see cref="T:string[]" /> containing the process command-line arguments.</param>
+        /// <param name="arguments">A <see cref="T:Tuple&lt;Dictionary&lt;string, string&gt;, FileInfo, string, Dictionary&lt;string, string&gt;&gt;" /> that receives the parsed command-line arguments.</param>
+        /// <returns><code>true</code> if the arguments were successfully parsed, otherwise <code>false</code>.</returns>
+        private static bool TryParseArguments(string[] args, out (Dictionary<string, string> Options, FileInfo MSBuildExePath, string EntryProjectPath, Dictionary<string, string> MSBuildGlobalProperties) arguments)
```
`MSBuildExeFile` and `EntryProjectFilePath`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,150 @@
+
+                return File.Exists(path) ? Assembly.LoadFrom(path) : null;
+            };
+#endif
```
Why is this assembly resolution conditional on `DEBUG`?  In general, I want to avoid differences between retail and debug code.  Otherwise, one can wind up in situations where something works on a local dev machine (debug by default) but not in a CI build (retail by default).

jeffkl replied:
When you build locally, the application base paths defined in the app.config won't work because they build output is not relative to the installed MSBuild.  So the DEBUG version (ie locally built) needs to have an assembly resolve event handler.  This does mean that if you build RELEASE locally, it won't work unless you copy your build output over the installed NuGet binaries. 

I could leave the assembly resolve event handler code in for both RELEASE and DEBUG and it wouldn't run if you were using an installed NuGet.  What do you prefer?

dtivel replied:
I'd prefer the same behavior for both release and debug configurations.  This block could be opt-in for both configurations.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ProjectItemInstanceEvaluatedIncludeComparer.cs:
```diff
@@ -0,0 +1,43 @@
+            return string.Equals(x?.Identity, y?.Identity, StringComparison.OrdinalIgnoreCase);
+        }
+
+        public int GetHashCode(IMSBuildItem obj) => StringComparer.OrdinalIgnoreCase.GetHashCode(obj.Identity.GetHashCode());
```
Shouldn't this be:
```C#
public int GetHashCode(IMSBuildItem obj) => StringComparer.OrdinalIgnoreCase.GetHashCode(obj.Identity);
```
Can you please add some unit tests around this class?

jeffkl replied:
Great catch, unit tests have been added

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,259 @@
+                        }
+                    }
+                    catch (Exception e) when (
+                        e is TaskCanceledException
```
`TaskCanceledException` subclasses `OperationCanceledException` so you don't need this condition.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,259 @@
+
+                        if (!process.HasExited)
+                        {
+                            process.Kill();
```
`Process.Kill()` can throw, such as during a race condition calling this and the process exiting.

Do [this](https://github.com/NuGet/NuGet.Client/blob/91d2331328d55ab9f450af1fbf184fc1f261ed25/src/NuGet.Core/NuGet.Credentials/PluginCredentialProvider.cs#L306-L322).

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,150 @@
+            try
+            {
+                var options = ParseSemicolonDelimitedListOfKeyValuePairs(args[0]);
+                var msbuildExePath = new FileInfo(args[1]);
```
`msbuildExeFile`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,150 @@
+            {
+                var options = ParseSemicolonDelimitedListOfKeyValuePairs(args[0]);
+                var msbuildExePath = new FileInfo(args[1]);
+                var entryProjectPath = args[2];
```
`entryProjectFilePath`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ProjectWithInnerNodes.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            if (string.IsNullOrWhiteSpace(targetFramework))
+            {
+                OuterProject = project;
```
`OuterProject` is not write-once and can be overwritten by successive calls to `Add(…)`.  Besides wanting to reuse code in `Add(…)`, is there a use case for allowing this?

For example:
```C#
var projectWithInnerNodes = new ProjectWithInnerNodes("tfm", outerProject1);

projectWithInnerNodes.Add(targetFramework: null, outerProject2);
```

jeffkl replied:
MSBuild can evaluate inner nodes before it evaluates outer nodes depending on your project-to-project references.  So this code keeps track of the inner nodes until finally the outer node is evaluated and set.  I don't think there's a situation where the outer node will actually be set twice, since MSBuild should not be evaluating the same project twice.

I could guard against it but the API I'm calling won't ever hit it and our API isn't public.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ProjectWithInnerNodes.cs:
```diff
@@ -0,0 +1,52 @@
+                TryAdd(targetFramework, project);
+            }
+
+            return this;
```
Why return anything at all?  The caller already has a reference to `this`.

jeffkl replied:
This is because in `DependencyGraphSpecGenerator`, I'm calling `AddOrUpdate` which expects a return value.  So this returns the collection itself so I can have less code in my AddOrUpdate delegate.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs:
```diff
@@ -283,9 +283,10 @@ public static void AddPropertyIfExists(IDictionary<string, string> properties, s
         /// <returns>A <see cref="Tuple{ProjectStyle, Boolean}"/> containing the project style and a value indicating if the project is using a style that is compatible with PackageReference.
         /// If the value of <paramref name="restoreProjectStyle"/> is not empty and could not be parsed, <code>null</code> is returned.</returns>
-        internal static (ProjectStyle ProjectStyle, bool IsPackageReferenceCompatibleProjectStyle) GetProjectRestoreStyle(string restoreProjectStyle, bool hasPackageReferenceItems, string projectJsonPath, string projectDirectory, string projectName, Common.ILogger log)
+        internal static (ProjectStyle ProjectStyle, bool IsPackageReferenceCompatibleProjectStyle, string PackagesConfigPath) GetProjectRestoreStyle(string restoreProjectStyle, bool hasPackageReferenceItems, string projectJsonPath, string projectDirectory, string projectName, Common.ILogger log)
```
`PackagesConfigFilePath`

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,115 @@
+        /// </summary>
+        public ConsoleOutLogMessageType MessageType { get; set; }
+
+        /// <inheritdoc cref="BuildMessageEventArgs.ColumnNumber" />
```
Fix this XML doc comment.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/LoggingQueue.cs:
```diff
@@ -0,0 +1,59 @@
+            return _queue.Post(obj);
+        }
+
+        public void Dispose()
```
This should be idempotent.  Is it?

jeffkl replied:
Yes, everything it calls is idempotent.

https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.actionblock-1.complete?view=netcore-3.1
https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.actionblock-1.completion?view=netcore-3.1

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,259 @@
+#endif
+            // Full path to the entry project.  If its a solution file, it will be the full path to solution, otherwise SolutionPath is either empty
+            // or is the value "*Undefined*" and ProjectFullPath is set instead.
+            yield return !string.IsNullOrWhiteSpace(SolutionPath) && !string.Equals(SolutionPath, "*Undefined*", StringComparison.OrdinalIgnoreCase)
```
Where does `*Undefined*` come from?

jeffkl replied:
https://github.com/microsoft/msbuild/blob/e812790eabece155f9ea7cbc13d57ce1ad7bff19/src/Tasks/Microsoft.Common.CurrentVersion.targets#L306-L324

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/RestoreTaskEx.cs:
```diff
@@ -0,0 +1,259 @@
+                            msBuildGlobalProperties.AddRange(globalProperties);
+                        }
+                    }
+                    catch
```
```C#
catch (Exception)
```

dtivel commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/MockMSBuildProject.cs:
```diff
@@ -0,0 +1,60 @@
+        }
+
+        public MockMSBuildProject(string fullPath)
+            : this(fullPath, null)
```
Add named parameter for `null`.

dtivel commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/MockMSBuildProject.cs:
```diff
@@ -0,0 +1,60 @@
+        }
+
+        public MockMSBuildProject(string fullPath, IDictionary<string, string> properties)
+            : this(fullPath, properties, null)
```
Add named parameter for `null`.

dtivel commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/MockMSBuildProject.cs:
```diff
@@ -0,0 +1,60 @@
+        public string FullPath { get; }
+
+        public IDictionary<string, IList<IMSBuildItem>> Items { get; set; }
+
```
Is a non-private setter necessary?

jeffkl replied:
Yes it allows tests to use the nicer initializer syntax rather than only call the constructor and pass in an object.

dtivel commented on test/TestUtilities/Test.Utility/MockSettings.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            OnSettingsChanged();
+
+            throw new NotImplementedException();
```
Why thrown after calling `OnSettingsChanged()`?  Why not just throw?

jeffkl replied:
Not sure, I'll remove

dtivel commented on test/TestUtilities/Test.Utility/MockSettings.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            OnSettingsChanged();
+
+            throw new NotImplementedException();
```
Why thrown after calling `OnSettingsChanged()`?

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks/ConsoleOutLogMessage.cs:
```diff
@@ -0,0 +1,115 @@
+    internal sealed class ConsoleOutLogMessage
+    {
+        /// <inheritdoc cref="BuildMessageEventArgs.Code" />
+        public string Code { get; set; }
```
Can all setters be removed?

jeffkl replied:
I'm using implicit operators instead of explicit constructors otherwise I could.  The class is internal and sealed so it probably doesn't matter unless you really prefer.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+        {
+            Debug = debug;
+
+            // TODO: Pass verbosity from main process
```
Either address this in this PR or create a work item but don't check in a TODO.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+        /// <summary>
+        /// Gets or sets a value indicating if this application is being debugged.
+        /// </summary>
+        public bool Debug { get; set; }
```
Remove setter.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+
+namespace NuGet.Build.Tasks.Console
+{
+    internal class DependencyGraphSpecGenerator : IDisposable
```
Can this be `sealed`?  If yes, make `protected` members `private`.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+        /// </summary>
+        protected MSBuildLogger MSBuildLogger => _msBuildLoggerLazy.Value;
+
+        public void Dispose()
```
Is this idempotent?

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+        /// <param name="options">The set of options to use when restoring.  These options come from the main MSBuild process and control how restore functions.</param>
+        /// <returns><code>true</code> if the restore succeeded, otherwise <code>false</code>.</returns>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public async Task<bool> RestoreAsync(string entryProjectPath, Dictionary<string, string> globalProperties, Dictionary<string, string> options)
```
`entryProjectFilePath`

Can the last 2 parameter types be `IReadOnlyDictionary<string, string>`?

jeffkl replied:
I can make the options object `IReadOnlyDictionary` but the `globalProperties` is passed to an MSBuild API that needs `IDictionary` (which is probably an oversight).

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+
+                        if (packageSpec != null)
+                        {
+                            // TODO: Make the backing collection Concurrent in future PR
```
Create work item instead of TODO.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+
+                sw.Stop();
+
+                // TODO: Localized resource
```
Address in this PR.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+                RuntimeGraph = new RuntimeGraph(
+                    MSBuildStringUtility.Split($"{project.GetProperty("RuntimeIdentifiers")};{project.GetProperty("RuntimeIdentifier")}")
+                        .Concat(projectsByTargetFramework.Values.SelectMany(i => MSBuildStringUtility.Split($"{i.GetProperty("RuntimeIdentifiers")};{i.GetProperty("RuntimeIdentifier")}")))
+                        .Distinct(StringComparer.Ordinal)
```
Runtime identifiers are opaque strings and therefore case-sensitive?

jeffkl replied:
Apparently not: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs#L507

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,896 @@
+
+                sw.Stop();
+
+                // TODO: Localized resource
```
Address in this PR.

jeffkl committed:
Address comments

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs:
```diff
@@ -0,0 +1,184 @@
+    {
+        /// <summary>
+        ///
+        /// </summary>
```
Remove or update.

dtivel commented on src/NuGet.Core/NuGet.Build.Tasks.Console/TaskLoggingQueueTests.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Delete this file.

jeffkl committed:
Address comments

jeffkl committed:
Address comments

nkolev92 commented:
@jeffkl I'll review this today or/and tomorrow :)

Just wanted to let oyu know before you merge :)

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/App.config:
```diff
@@ -0,0 +1,62 @@
+    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+      <!-- Use codeBase entries for MSBuild assemblies so that the NGen'd native images will be used -->
+      <dependentAssembly>
+        <assemblyIdentity name="Microsoft.Build.Framework" culture="neutral" publicKeyToken="b03f5f7f11d50a3a" />
```
Can you elaborate why we need all these binding redirects?

jeffkl replied:
Two reasons:

* Tasks loaded by MSBuild might have been compiled against older versions of MSBuild.  Since we're executing MSBuild 15+, these binding redirects make sure that user task assemblies will load properly even if they are compiled against older versions of MSBuild.
* We cannot ship MSBuild next to the application, so we need a way to locate it.  If we use assembly binding at runtime, the CLR will not choose the NGEN'd assemblies and JIT times are much slower.  Setting the basePaths to relative locations next to NuGet lets us load the NGen'd assemblies.

nkolev92 replied:
Thanks, I'd be curious to understand how we can keep things in sync, but sounds you already pinged Rainer :)

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+        {
+            try
+            {
+                // TODO: Use a localized resource from https://github.com/NuGet/NuGet.Client/pull/3111
```
Each TODO should have a tracking issue. 
Or you can resolve it and fix it up later when that other PR is ready.

jeffkl replied:
I was trying to avoid having to handle the merge conflict later.  Am I expected to open issues for code that hasn't been merged yet?  My plan was as part of the other PR to update this code path to use the new localized resource.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/ConsoleLoggingQueue.cs:
```diff
@@ -0,0 +1,181 @@
+#if IS_CORECLR
+using Microsoft.Build.Utilities;
+#endif
+using Newtonsoft.Json;
```
Fix the imports. 
You don't need Newtonsoft.Json.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+using NuGet.ProjectModel;
+using NuGet.RuntimeModel;
+using NuGet.Versioning;
+using ILogger = Microsoft.Build.Framework.ILogger;
```
nit: Should we use a different name? Especially if the type is defined on the nuget side as well. 

For example, IMSBuildLogger.

jeffkl replied:
In this case, its only used in one place so I'll just use the fully qualified name in that spot.
```C#
var loggers = new List<Microsoft.Build.Framework.ILogger>
{
    LoggingQueue
};
```

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+                {
+                    // Use the same loggers as the project collection
+                    Loggers = projectCollection.Loggers,
+                    LogTaskInputs = Debug
```
This is more for my own curiosity than anything else. 
At what log level are the task inputs logged? 

Basically my thinking is I shouldn't need to have debug bits, to get these detailed logs. 
Not sure if it'll ever be useful, but just curious :)

jeffkl replied:
Task inputs are logged at the detailed level.  But I've set it up so you can either debug the app or set the env var `RESTORE_TASK_BINLOG_PARAMETERS` and you'll get a full binary log with task inputs. Users will set the env var for troubleshooting.

nkolev92 replied:
Makes sense.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+        /// <param name="options">The set of options to use when restoring.  These options come from the main MSBuild process and control how restore functions.</param>
+        /// <returns><code>true</code> if the restore succeeded, otherwise <code>false</code>.</returns>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public async Task<bool> RestoreAsync(string entryProjectFilePath, IDictionary<string, string> globalProperties, IReadOnlyDictionary<string, string> options)
```
Feels like this class shouldn't actually be doing the restore. 

The name suggests it just generates the DG Spec. 

I'd either change the name or just separate them. I imagine you kept them in one class because in here you define the logging queue.

jeffkl replied:
I'll rename it to something like MSBuildStaticGraphRestore

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+        }
+
+        /// <summary>
+        /// Gets the framework references for the specified project.
```
Can you please clarify the ones that are per framework (almost everything is). 

We know these, but feel like it'd be helpful for a future reader.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+            }
+
+            // If the project did not specify a value for TargetFrameworks, return the short folder name of the frameworks which where parsed via other properties
+            var targetFrameworks = new List<string>(frameworks.Count);
```
I think this part is incorrect. 

See: https://github.com/NuGet/NuGet.Client/blob/feceff9099e6f02dd2a6d76b26ed68a6029959f5/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs#L798

We only set the original target frameworks when there's multi targeting involved. 
Otherwise we might end up adding an unnecessary condition. 

It might break no-op. 

We should ensure we add no-op tests similar to the ones that got broken by the other PR :)

jeffkl replied:
The `TargetFrameworks` property isn't read from the project itself, but the ProjectSpec item in the restore graph.  That is populated in this target:
https://github.com/NuGet/NuGet.Client/blob/feceff9099e6f02dd2a6d76b26ed68a6029959f5/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L479

By this task:

https://github.com/NuGet/NuGet.Client/blob/feceff9099e6f02dd2a6d76b26ed68a6029959f5/src/NuGet.Core/NuGet.Build.Tasks/GetProjectTargetFrameworksTask.cs#L71

And assigned to the ProjectSpec item here: https://github.com/NuGet/NuGet.Client/blob/feceff9099e6f02dd2a6d76b26ed68a6029959f5/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L693

My code starts with the same logic: https://github.com/NuGet/NuGet.Client/blob/c8661279a93886e4a2554dbd538f39ef14433425/src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs#L342

I restored with and without my changes and the assets and dgspec files are identical so I'm confident it does not break no-op

nkolev92 replied:
Ah, I missed that. Thanks for the explanation. 

These are sometimes hard to follow cause we go from targets => task => targets => task to all task. 

In the future, it should be easier :)

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+                foreach (var version in MSBuildStringUtility.Split(projectItemInstance.GetProperty("Version")))
+                {
+                    // Validate the version range
+                    VersionRange versionRange = !string.IsNullOrWhiteSpace(version) ? VersionRange.Parse(version) : VersionRange.All;
```
We should create a task to consider refactoring these into shared utilities as necessary. 
The simpler subroutines cause the larger ones are a bit more difficult (we have ones for restore settings for example which consider the configs).

I remember we consciously opted for this because of the scope creep.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+
+                libraryDependencies.Add(new LibraryDependency
+                {
+                    AutoReferenced = packageReferenceItem.IsPropertyTrue("IsImplicitlyDefined"),
```
Can you please define an internal utility with the string names. 
I trust there probably isn't one, but seeing this makes me scream typo! 

MSBuildRestoreUtility is broken that way. this doesn't need to be. 

Similar to the above, a follow-up issue might be ok, but I'd want to make sure we address it before shipping.

nkolev92 replied:
Maybe that task should also consider https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectBuildProperties.cs. 

it already contains some properties, but currently they are only used in VS, maybe it can be moved down a level :)

jeffkl replied:
I don't currently reference `NuGet.PackageManagement` so I couldn't use these consts.  I'd love to clean this up in a later PR but I really need to just get this in.

nkolev92 replied:
Yeah, we shouldn't reference NuGet.PackageManagement...I'd rather just duplicate it down, but yeah better handled in a future PR.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+
+            foreach (var projectReferenceItem in projectReferenceItems)
+            {
+                string fullPath = projectReferenceItem.GetProperty("FullPath");
```
Shouldn't it be MSBuildProjectFullPath?

jeffkl replied:
`FullPath` is special metadata that returns the full path for an item
https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-well-known-item-metadata?view=vs-2019
`MSBuildProjectFullPath` is a property for the project, but we need the item's metadata here.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+        internal static List<ProjectRestoreReference> GetProjectReferences(IMSBuildProject project)
+        {
+            // Get the unique list of ProjectReference items that have the ReferenceOutputAssembly metadata set to "true", ignoring duplicates
+            var projectReferenceItems = project.GetItems("ProjectReference")
```
Make sure you don't have a casing problem here. 

See https://github.com/NuGet/NuGet.Client/blob/feceff9099e6f02dd2a6d76b26ed68a6029959f5/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs#L351. 

We need to have consistent casing throughout the dg spec.

See the scenario here: https://github.com/NuGet/Home/issues/5855

nkolev92 replied:
In general we have a handling "attempt" in the old restore codepath that doesn't seem accounted for here. 

I'm not sure the old approach does much tbh, given that MSBuild treats all items case insensitively, but the lack of casing normalization might resurrect the old bug.

jeffkl replied:
Great catch, I'll update my logic

nkolev92 replied:
The change looks good, just make sure we add a test for this scenario. 

The challenge with all this work is that most of our restore functionality tests are nuget.exe based cause our infrastructure is well built for that. 

With this change, the nuget.exe and msbuild restore codepaths diverge quite a bit.

nkolev92 replied:
Create an issue for this test gap. https://github.com/NuGet/Home/issues/9037
If possible, please make sure to add a test for the scenario in the tests you will add in a future PR.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+                        project.Directory,
+                        project.SplitPropertyValueOrNull("RestoreFallbackFolders"),
+                        project.SplitPropertyValueOrNull("RestoreFallbackFoldersOverride"),
+                        innerNodes.SelectMany(i => MSBuildStringUtility.Split(i.GetProperty("RestoreAdditionalProjectFallbackFolders"))),
```
We should make sure we have tests for this.

jeffkl replied:
Future end-to-end tests will cover this

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+            var projectStyleResult = BuildTasksUtility.GetProjectRestoreStyle(
+                restoreProjectStyle: project.GetProperty("RestoreProjectStyle"),
+                hasPackageReferenceItems: targetFrameworkInfos.Any(i => i.Dependencies.Any()),
+                projectJsonPath: project.GetProperty("_CurrentProjectJsonPath"),
```
Is project.json supported with this still/ 

Can you elaborate how this is supposed to work?

jeffkl replied:
I don't think it will work at the moment.  But I pass this along for now.  I was too focused on getting this working at all to spend time on project.json.  In a future PR I can log an error or I can try and get it working.

nkolev92 replied:
I don't necessarily think we want to care about project.json much, but we should get clarity on it. 
If you ask me, I want it gone by 6.0 :D 

https://github.com/NuGet/Home/issues/7199

Just worth tracking/calling out.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/DependencyGraphSpecGenerator.cs:
```diff
@@ -0,0 +1,893 @@
+                            EvaluationContext = evaluationContext,
+                            GlobalProperties = properties,
+                            // Ignore bad imports to maximize the chances of being able to load the project and restore
+                            LoadSettings = ProjectLoadSettings.IgnoreEmptyImports | ProjectLoadSettings.IgnoreInvalidImports | ProjectLoadSettings.IgnoreMissingImports | ProjectLoadSettings.DoNotEvaluateElementsWithFalseCondition,
```
This is a change from current restore. 
ProjectLoadSettings.IgnoreMissingImports is not the default in current restore. 

Granted the move was for performance reasons (in 15.5), but it changed the experience for non-SDK based projects. 
There'll be a difference in the experience :)

nkolev92 replied:
fwiw, I get the arguments for having it, but we try to avoid having different experiences across client, because it just adds cognitive overload for both us and customers trying to figure out how the heck things are supposed to work, or why they work in client but break in another.

jeffkl replied:
I believe its already this way for `msbuild /t:restore` and Visual Studio since they both set these flags.  So its only different from `nuget.exe restore` but we're working to deprecate that.

nkolev92 replied:
I'm certain that msbuild /t:restore doesn't, because people have hit it before :)
Visual Studio won't load projects whose imports cannot be resolved for the most part.

jeffkl replied:
Here is where MSBuild sets the flags: https://github.com/microsoft/msbuild/blob/master/src/MSBuild/XMake.cs#L1356

nkolev92 replied:
Maybe the effect of these is not actually the one that I am thinking of. 

I have forwarded you a thread about that, to help clarify it.

nkolev92 replied:
All of my confusion is summarized in https://github.com/NuGet/Home/issues/9035. 

No regression, this change is not relevant for this PR.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/NuGet.Build.Tasks.Console.csproj:
```diff
@@ -0,0 +1,55 @@
+    <NoWarn>$(NoWarn);CS1591;NU5100</NoWarn>
+    <XPLATProject>true</XPLATProject>
+    <EnableDefaultNoneItems>false</EnableDefaultNoneItems>
+    <ServerGarbageCollection>true</ServerGarbageCollection>
```
Did we run this by the GC team?
Curious what their response was?

jeffkl replied:
I have not had time to ask around.  I'm confident this is the correct choice.

nkolev92 replied:
I think it's ok as well, just curious if they had an opinion :)

nkolev92 replied:
Well, an opinion more than the readily available docs.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/NuGet.Build.Tasks.Console.csproj:
```diff
@@ -0,0 +1,55 @@
+
+  <ItemGroup Condition="$(DefineConstants.Contains(SIGNED_BUILD))">
+    <AssemblyAttribute Include="System.Runtime.CompilerServices.InternalsVisibleTo">
+      <_Parameter1>NuGet.Build.Tasks.Test, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9</_Parameter1>
```
The project should get its own test assembly, not reuse an old one.

jeffkl replied:
I'd like to fold the assemblies together in the next release personally so I reused the test assembly.  The two share so much logic I thought it made sense.  Its only a separate app because I wasn't allowed to rename the task DLL to an EXE.

nkolev92 replied:
I think I'd favor a new assembly, but I'm curious what the rest of the team thinks @NuGet/nuget-client. 

Let's not block this PR on this. I will ping them in teams.

zivkan replied:
My 2c. I do generally prefer a 1:1 mapping between production and test projects. This makes so when someone is modifying/reading the production code, it's easy to find the test code. Just look for the test project with the same prefix as the prod project's name.

If the two projects share logic, I hope that means it's in one place and referenced from the other, not duplicated. In which case it can be tested in the one place it's written. I didn't review this PR in detail though, so perhaps I'm missing something.

dominoFire replied:
I prefer a separate test assembly for testing code that lives in `NuGet.Build.Tasks.Console`. That makes easy to navigate between test and production code, as @zivkan pointed out.

Also, we can address this in a later PR. If you're not mixing `NuGet.Build.Tasks` tests that run with `NuGet.Build.Tasks.Console`, I think we are good to go.

dominoFire replied:
For the folding assemblies case, we have tests for `NuGet.CommandLine` (nuget.exe) and for other library assemblies as well (`NuGet.Commands`, `NuGet.Protocol`, etc.)

jeffkl replied:
Okay I'll make a new test assembly as a follow-up PR

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,154 @@
+            }
+
+            // Parse command-line arguments
+            if (!TryParseArguments(args, out var arguments))
```
Don't use var, the type is not obvious.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs:
```diff
@@ -0,0 +1,154 @@
+                };
+            }
+
+            using (var dependencyGraphSpecGenerator = new DependencyGraphSpecGenerator(debug: debug))
```
I think the generator should ideally just generate the dg spec, and then you can call build.task.utility to restore from here.

jeffkl replied:
`Program` can't load any MSBuild types or the assembly resolution won't work when you're debugging a locally built copy of the app.  I'll just rename my class.

nkolev92 replied:
I see. Should we add a quick code comment, cause that seems like something that tests won't catch, but it could affect our debugging flow.

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.Build.Tasks.csproj:
```diff
@@ -40,6 +46,10 @@
   </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="System.Threading.Tasks.Dataflow" Version="4.9.0" />
```
Where is this used?

jeffkl replied:
It's used by the logging queue: https://github.com/NuGet/NuGet.Client/pull/3109/files#diff-4b17c63a81b794616ee9ad377f6d5c80R5

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.Build.Tasks.csproj:
```diff
@@ -56,6 +66,24 @@
-  <Import Project="$(BuildCommonDirectory)common.targets"/>
+  <ItemGroup Condition="$(DefineConstants.Contains(SIGNED_BUILD))">
+    <AssemblyAttribute Include="System.Runtime.CompilerServices.InternalsVisibleTo">
+      <_Parameter1>NuGet.Build.Tasks.Console, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9</_Parameter1>
```
We don't use InternalsVisibleTo between [shipping assemblies](https://github.com/NuGet/NuGet.Client/pull/3180/files#diff-85d4cc615fcaaf78685c6435fc0f6de2R158)

jeffkl replied:
Okay I'll make the classes public

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.RestoreEx.targets:
```diff
@@ -0,0 +1,24 @@
+        ForceEvaluate="$(RestoreForceEvaluate)"
+        HideWarningsAndErrors="$(HideWarningsAndErrors)"
+        IgnoreFailedSources="$(RestoreIgnoreFailedSources)"
+        Interactive="$([MSBuild]::ValueOrDefault('$(NuGetInteractive)', '$(MSBuildInteractive)'))"
```
neat :D 
Has this always been there/

jeffkl replied:
I think its been around since MSBuild 12
https://docs.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2019#msbuild-valueordefault

nkolev92 commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -1259,4 +1259,15 @@ Copyright (c) .NET Foundation. All rights reserved.
+    won't exist in that case.
+    ============================================================
+  -->
+  <Import Project="NuGet.RestoreEx.targets" Condition="'$(RestoreEnableStaticGraph)' == 'true' And Exists('NuGet.RestoreEx.targets')" />
```
nit: RestoreEnableStaticGraph => RestoreUseStaticGraphEvaluation maybe? 

Potentially in line with other properties :)

jeffkl replied:
Done

nkolev92 reviewed (commented):
Looks great overall. 

A bunch of smaller tasks that need clarified.

Do we have a spec for this anywhere?

jeffkl committed:
Address comments

jeffkl commented:
@nkolev92 please resolve conversations if I addressed your question and let me know what else you want me to change so we can get this merged.

dtivel commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/DependencyGraphSpecGeneratorTests.cs:
```diff
@@ -0,0 +1,705 @@
+
+namespace NuGet.Build.Tasks.Test
+{
+    public class DependencyGraphSpecGeneratorTests
```
Rename class and file name to MSBuildStaticGraphRestoreTests

dtivel commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Test/DependencyGraphSpecGeneratorTests.cs:
```diff
@@ -0,0 +1,705 @@
+        [InlineData("(1.2.3,]")]
+        [InlineData("1.*")]
+        [InlineData("[1.2.3];4.5.6", "4.5.6")]
+        public void DependencyGraphSpecGenerator_GetPackageDownloads_ThrowsWhenNotExactVersion(string version, string expected = null)
```
Rename tests:

    <MemberName>_<Condition>__<ExpectedResult>

This would be:

    GetPackageDownloads_WhenNotExactVersion_Throws

The `DependencyGraphSpecGenerator` name is obsolete.

jeffkl committed:
Address comments

nkolev92 commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -348,7 +348,7 @@ public static void RemoveMissingProjects(DependencyGraphSpec graphSpec)
         /// Change all project paths to the same casing.
         /// </summary>
-        public static void NormalizePathCasings(Dictionary<string, string> paths, DependencyGraphSpec graphSpec)
+        public static void NormalizePathCasings(IDictionary<string, string> paths, DependencyGraphSpec graphSpec)
```
I don't know if anyone is using, but I think we should the breaking change.

jeffkl committed:
Avoid breaking change

nkolev92 reviewed (approved):
I think the only leftover comment is create a new assembly for the tests. 

The rest of the issues were either addressed/will be a added in a future PR/have a follow issue.

jeffkl commented:
@nkolev92 CI appears to be broken again, how can I get a green build?

zivkan commented:
#3184 should fix the most common Apex & E2E test failures. It was merged a few hours ago, so please rebase and try again.

ViktorHofer commented:
@jeffkl I was trying to use RestoreTaskEx but in my sdk (5.0.100-preview.2.20155.14) the `NuGet.Build.Tasks.Console.dll` isn't present. Also it seems that errors are swallowed as that command invoked the RestoreTaskEx task but no output was logged and no assets files were generated, this was basically a no-op:

```
C:\git\runtime4\tools-local\tasks\installer.tasks>..\..\..\dotnet.cmd msbuild /t:Restore /p:RestoreUseStaticGraphEvaluation=true

Microsoft (R) Build Engine version 16.6.0-preview-20105-01+5d872c945 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.


C:\git\runtime4\tools-local\tasks\installer.tasks>
```

Am I missing something here? cc @nkolev92

ViktorHofer commented:
OK found https://github.com/NuGet/Home/issues/9267. Still we should follow-up why errors aren't propagated.

nkolev92 commented:
@ViktorHofer Hey, can you please file an issue for the error propagation?

ViktorHofer commented:
Sure: https://github.com/NuGet/Home/issues/9282
